### PR TITLE
Make gonzobot base instead of cloudbot

### PR DESCRIPTION
**Pull request by tuzonghua**

### CODE REVIEW

**Comment by tuzonghua** [comment]
@weylin @danburbul As this is a very large change, I would suggest looking at only a few files. Namely, all the configs (config*.json) and any plugins that we've made edits to.

File: README.md [hunk 1]
```diff
@@ -49,7 +51,7 @@ More at the [Wiki Main Page](https://github.com/CloudBotIRC/CloudBot/wiki).
 
 The developers reside in [#CloudBot](irc://irc.esper.net/cloudbot) on [EsperNet](http://esper.net) and would be glad to help you.
 
-If you think you have found a bug/have an idea/suggestion, please **open an issue** here on Github and contact us on IRC!
+If you think you have found a bug/have a idea/suggestion, please **open a issue** here on Github and contact us on IRC!
```
Comment (JacobCallahan) [inline README.md:54]: are you sure about this?

File: cloudbot/client.py [hunk 2]
```diff
@@ -55,11 +69,33 @@ def __init__(self, bot, name, nick, *, channels=None, config=None):
...
+        yield from self.try_connect()
+
+    @asyncio.coroutine
+    def try_connect(self):
+        timeout = 30
+        while not self.connected:
+            try:
+                yield from self.connect(timeout)
+            except Exception:
+                logger.exception("[%s] Error occurred while connecting.", self.name)
```
Comment (JacobCallahan) [inline cloudbot/client.py:91]: Why not use ```.format``` here?

File: cloudbot/clients/irc.py [hunk 3]
```diff
@@ -42,38 +45,36 @@ def decode(bytestring):
...
         """
         super().__init__(bot, name, nick, channels=channels, config=config)
 
-        self.use_ssl = use_ssl
-        self._ignore_cert_errors = ignore_cert_errors
-        self._timeout = timeout
-        self.server = server
-        self.port = port
+        self.use_ssl = config['connection'].get('ssl', False)
+        self._ignore_cert_errors = config['connection']['ignore_cert']
```
Comment (JacobCallahan) [inline cloudbot/clients/irc.py:69]: is there a sensible default you can add here?

File: cloudbot/clients/irc.py [hunk 4]
```diff
@@ -103,66 +99,106 @@ def describe_server(self):
...
-        if self._connected:
+        yield from self.try_connect()
+
+    @asyncio.coroutine
+    def try_connect(self):
+        while not self.connected:
+            try:
+                yield from self.connect(self._timeout)
+            except (asyncio.TimeoutError, OSError):
+                logger.exception("[%s] Error occurred while connecting", self.name)
```
Comment (JacobCallahan) [inline cloudbot/clients/irc.py:121]: same question about using ```.format```

File: cloudbot/clients/irc.py [hunk 5]
```diff
@@ -103,66 +99,106 @@ def describe_server(self):
...
-            self.cmd("QUIT")
+        self._active = False
 
-    def close(self):
-        if not self._quit:
-            self.quit()
-        if not self._connected:
-            return
+        if self.connected:
+            if reason:
```
Comment (JacobCallahan) [inline cloudbot/clients/irc.py:177]: maybe ```self.cmd("QUIT", reason)``` would work, as long as reason is None i would hope the ```self.cmd``` wouldn't take action on it.

File: cloudbot/clients/irc.py [hunk 6]
```diff
@@ -103,66 +99,106 @@ def describe_server(self):
...
-    def action(self, target, text, sanatize=True):
-        if sanatize == True:
-            text = "".join(text.splitlines())
+    def admin_log(self, text, console=True):
+        log_chan = self.config.get("log_channel")
+        if log_chan:
+            self.message(log_chan, text)
+
+        if console:
+            logger.info("[%s|admin] %s", self.name, text)
```
Comment (JacobCallahan) [inline cloudbot/clients/irc.py:196]: same ```.format``` question

File: cloudbot/clients/irc.py [hunk 7]
```diff
@@ -255,46 +294,87 @@ def __init__(self, conn):
...
+
+        for out_sieve in self.bot.plugin_manager.out_sieves:
+            event = IrcOutEvent(
+                bot=self.bot, hook=out_sieve, conn=self.conn, irc_raw=line
+            )
+
+            ok, new_line = yield from self.bot.plugin_manager.internal_launch(out_sieve, event)
+            if not ok:
+                logger.warning("Error occurred in outgoing sieve, falling back to old behavior")
+                logger.debug("Line was: %s", line)
```
Comment (JacobCallahan) [inline cloudbot/clients/irc.py:354]: format?

File: cloudbot/clients/irc.py [hunk 8]
```diff
@@ -303,63 +383,32 @@ def data_received(self, data):
...
-                user = None
-                host = None
-                mask = None
-
-            command_params = irc_param_re.findall(params)
+            try:
+                message = Message.parse(line)
+            except Exception:
+                logger.exception(
+                    "[%s] Error occurred while parsing IRC line '%s' from %s",
```
Comment (JacobCallahan) [inline cloudbot/clients/irc.py:390]: format? gonna stop commenting on format. if it's not a bad idea, just search for all instances.

File: cloudbot/clients/irc.py [hunk 9]
```diff
@@ -392,22 +441,42 @@ def data_received(self, data):
...
+            nick = prefix.nick
+            user = prefix.user
+            host = prefix.host
+            mask = prefix.mask
+
+            if channel:
+                # TODO Migrate plugins to accept the original case of the channel
+                channel = channel.lower()
+
+                channel = channel.split()[0]  # Just in case there is more data
```
Comment (JacobCallahan) [inline cloudbot/clients/irc.py:464]: these lines can be combined into ```channel = channel.lower().split()[0]```

File: cloudbot/hook.py [hunk 10]
```diff
@@ -339,3 +377,126 @@ def _on_start_hook(func):
...
+
+def on_cap_available(*caps, **kwargs):
+    """External on_cap_available decorator. Must be used as a function that returns a decorator
+
+    This hook will fire for each capability in a `CAP LS` response from the server
+    """
+
+    def _on_cap_available_hook(func):
+        hook = _get_hook(func, "on_cap_available")
+        if hook is None:
```
Comment (JacobCallahan) [inline cloudbot/hook.py:412]: ```if not hook:``` ? 
there are a number of instances of this as well

File: cloudbot/plugin.py [hunk 11]
```diff
@@ -108,38 +123,50 @@ def load_all(self, plugin_dir):
...
-        :type path: str
+        :type path: str | Path
         """
 
-        file_path = os.path.abspath(path)
-        file_name = os.path.basename(path)
-        title = os.path.splitext(file_name)[0]
+        path = Path(path)
+        file_path = path.resolve()
+        file_name = file_path.name
```
Comment (JacobCallahan) [inline cloudbot/plugin.py:151]: this var is only used once, maybe just scrap this and use the ```.name``` method later on

File: plugins/badwords.py [hunk 12]
```diff
@@ -1,153 +1,102 @@
...
-    if not (text.split(' ')[1] or text.split(' ')[1]('#')):
-        return "please specify a valid channel name"
-    channel = text.split(' ')[1].lower()
-    db.execute(
-        "delete from badwords where word = :word and chan = :chan", {
-            "word": word, "chan": channel})
+            word, channel
+        )
+
+    if len(badcache[channel]) >= 10:
```
Comment (JacobCallahan) [inline plugins/badwords.py:53]: why the limit?

File: plugins/badwords.py [hunk 13]
```diff
@@ -1,153 +1,102 @@
...
-    load_bad(db, conn)
+    load_bad(db)
+    wordlist = list_bad(channel)
+    return "Current badwords: {}".format(wordlist)
+
+
+@hook.command("rmbad", "delbad", permissions=["badwords"])
+def del_bad(text, db):
+    """<word> <channel> - removes the specified word from the specified channels bad word list"""
+    splt = text.lower().split(None, 1)
```
Comment (JacobCallahan) [inline plugins/badwords.py:68]: ```word, channel = text.lower().split(None, 1)``` ?

File: cloudbot/client.py (hunk 2 above)
Comment (JacobCallahan) [inline cloudbot/client.py:91]: There are a number of instances of using this string formatting instead of ```.format```. i'd recommend changing it.

**Review by JacobCallahan — DISMISSED** [review]
A forced page reload removed a decent number of my additional comments so i'll have to go through it again when i get time.

**Review by JacobCallahan — APPROVED** [review]
ACK

**Pull request merged**